### PR TITLE
SliceInferer for slice-by-slice infererence

### LINKED ISSUE #3458: Extension for cross-dimensional inference with SlidingWindowInferer
**Is your feature request related to a problem? Please describe.**
The request is to have an addition for the SlidingWindowInferer to support inference on, for example, 3D CT scans on a slice-by-slice basis using a 2D model. 

**Describe the solution you'd like**
A simple solution exists ofcourse, where the model is run using SimpleInferer on a slice-by-slice basis and then collected. However, adding this support to SlidingWindowInferer will allow it to be more generalized. A single interface can handle the inference for a 3D volume both with 2D and 3D models. 

**Describe alternatives you've considered**
I've implemented a naive version here: https://github.com/ganslate-team/ganslate/blob/master/ganslate/utils/sliding_window_inferer.py
which looks at the `roi_size` and the `inputs` and adds a `network_wrapper` over the 2D model

**Additional context**
I would be happy to contribute a better solution if this is interesting to a larger set of user.

### PR DESCRIPTION
**Pull request by surajpaib**
Fixes #3458 

### Description
Addition of SliceInferer to the monai.inferers for slice-by-slice inference on a 3D volume using a 2D model. 

### Status
**Work in progress**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by surajpaib** [comment]
Hi @Nic-Ma, this is the first version of incorporating the `SliceInferer` into `monai.inferers`.

What do you think of the implementation? For now, it is the same as what was in the tutorials with better documentation. I'm trying to think of how to make it more readable and concise. Any inputs would be appreciated.

File: docs/source/inferers.rst [hunk 1]
```diff
@@ -36,3 +36,9 @@ Inferers
 .. autoclass:: SaliencyInferer
     :members:
     :special-members: __call__
+
+`SliceInferer`
+~~~~~~~~~~~~~~~~~
```
Comment (Nic-Ma) [inline docs/source/inferers.rst:41]: Please make the `~~~~` length same as `SliceInferer`.

Thanks.

File: monai/inferers/inferer.py [hunk 2]
```diff
@@ -221,3 +221,68 @@ def __call__(self, inputs: torch.Tensor, network: nn.Module, *args: Any, **kwarg
...
+
+class SliceInferer(SlidingWindowInferer):
+    """
+    SliceInferer extends SlidingWindowInferer to provide slice-by-slice (2D) inference
+    when provided a 3D volume.
+
+    Args:
+        spatial_dim: Spatial dimension over which the slice-by-slice inference runs on the 3D volume.
+            For example ``0`` could slide over axial slices. ``1`` over coronal slices and ``2`` over sagittal slices.
+        args: other optional args to be passed to the `__init__` of SliceInferer.
```
Comment (Nic-Ma) [inline monai/inferers/inferer.py:234]: I think it should be: "passed to the `__init__` of base class SlidingWindowInferer".

Thanks.

File: monai/inferers/inferer.py [hunk 3]
```diff
@@ -221,3 +221,68 @@ def __call__(self, inputs: torch.Tensor, network: nn.Module, *args: Any, **kwarg
...
+        kwargs: other optional keyword args to be passed to `__init__` of SliceInferer.
```
Comment (Nic-Ma) [inline monai/inferers/inferer.py:235]: I think it should be: "passed to the `__init__` of base class SlidingWindowInferer".

Thanks.

File: monai/inferers/inferer.py [hunk 4]
```diff
@@ -221,3 +221,68 @@ def __call__(self, inputs: torch.Tensor, network: nn.Module, *args: Any, **kwarg
...
+
+
+    ``args``, ``kwargs`` follow :py:class:`monai.inferer.SlidingWindowInferer`.
```
Comment (Nic-Ma) [inline monai/inferers/inferer.py:238]: After changing the `args` and `kwargs`, no need this line anymore.

Thanks.

File: monai/inferers/inferer.py [hunk 5]
```diff
@@ -221,3 +221,68 @@ def __call__(self, inputs: torch.Tensor, network: nn.Module, *args: Any, **kwarg
...
+        self, inputs: torch.Tensor, network: Callable[..., torch.Tensor], *args: Any, **kwargs: Any
+    ) -> torch.Tensor:
+        """
+        Args:
+            inputs: 3D input for inference
+            network: 2D model to execute inference on slices in the 3D input
+            args: optional args to be passed to ``network``.
+            kwargs: optional keyword args to be passed to ``network``.
+        """
+        assert self.spatial_dim < 3, "`spatial_dim` can only be `[D, H, W]` with `0, 1, 2` respectively"
```
Comment (Nic-Ma) [inline monai/inferers/inferer.py:256]: Please change to use:
```py
if self.spatial_dim < 3:
    raise ValueError("`spatial_dim` can only be `[H, W, D]` with `0, 1, 2` respectively.")
```

File: monai/inferers/inferer.py [hunk 6]
```diff
@@ -221,3 +221,68 @@ def __call__(self, inputs: torch.Tensor, network: nn.Module, *args: Any, **kwarg
...
+
+        # Check if roi size (eg. 2D roi) and input volume sizes (3D input) mismatch
+        if len(self.roi_size) != len(inputs.shape[2:]):
```
Comment (Nic-Ma) [inline monai/inferers/inferer.py:259]: Could you please help add some doc-string to describe the expected `roi_size`?

Thanks.

File: monai/inferers/inferer.py [hunk 7]
```diff
@@ -221,3 +221,68 @@ def __call__(self, inputs: torch.Tensor, network: nn.Module, *args: Any, **kwarg
...
+
+            # If they mismatch and roi_size is 2D add another dimension to roi size
+            if len(self.roi_size) == 2:
+                self.roi_size = list(self.roi_size)
+                self.roi_size.insert(self.spatial_dim, 1)
+            else:
+                raise RuntimeError("Currently, only 2D `roi_size` is supported, cannot broadcast to volume. ")
+
+        return super().__call__(inputs, lambda x: self.network_wrapper(network, x))
```
Comment (Nic-Ma) [inline monai/inferers/inferer.py:268]: Seems missed to pass the `args` and `kwargs` to `network_wrapper`.

Thanks.

**Review by Nic-Ma — COMMENTED** [review]
Thanks for the quick implementation!
Put some minor comments inline.
And please add some **unit tests** to cover the logic.

Thanks.

**Comment by surajpaib** [comment]
Hi @Nic-Ma 

Incorporated all the changes you suggested and added a unit test. Let me know if something is missing. 

I also changed the logic a bit to make the purpose of the `SliceInferer` more clear. The old version would fallback and run the `SlidingWindowInferer` if the `roi_size` was not provided according to the `SliceInferer` requirements - this might be confusing to the user, therefore I've changed it to throw a `RuntimeError` in case the `roi_size` is not as expected. Added a note to describe what `roi_size` expects too.

File: monai/inferers/inferer.py [hunk 8]
```diff
@@ -222,3 +222,60 @@ def __call__(self, inputs: torch.Tensor, network: nn.Module, *args: Any, **kwarg
...
+    ) -> torch.Tensor:
+        """
+        Args:
+            inputs: 3D input for inference
+            network: 2D model to execute inference on slices in the 3D input
+            args: optional args to be passed to ``network``.
+            kwargs: optional keyword args to be passed to ``network``.
+        """
+        if self.spatial_dim > 2:
+            raise ValueError("`spatial_dim` can only be `[H, W, D]` with `0, 1, 2` respectively.")
```
Comment (Nic-Ma) [inline monai/inferers/inferer.py:259]: I feel maybe the message should be "`spatial_dim` can only be `[0, 1, 2]` with `[H, W, D]` respectively."?

Thanks.

**Review by Nic-Ma — APPROVED** [review]
Hi @surajpaib ,

Really sorry for the late response, I missed this PR last week..
Looks good to me, put minor comments inline.
@wyli Could you please help double confirm it?

Thanks.

**Comment by Nic-Ma** [comment]
Hi @surajpaib ,

There is some format error in the code according to CI tests:
```
monai/inferers/inferer.py:262:16: error: Argument 1 to "len" has incompatible type "Union[Sequence[int], int]"; expected "Sized"  [arg-type]
monai/inferers/inferer.py:263:34: error: Argument 1 to "list" has incompatible type "Union[Sequence[int], int]"; expected "Iterable[int]"  [arg-type]
monai/inferers/inferer.py:268:41: error: Argument 2 to "__call__" of "SlidingWindowInferer" has incompatible type "Callable[[Any], Callable[..., Tensor]]"; expected "Callable[..., Tensor]"  [arg-type]
monai/inferers/inferer.py:268:51: error: Incompatible return value type (got "Callable[..., Tensor]", expected "Tensor")  [return-value]
monai/inferers/inferer.py:281:16: error: Incompatible return value type (got "Tensor", expected "Callable[..., Tensor]")  [return-value]
```

Thanks.

**Pull request merged**